File: apps/api/src/modules/reviews/mappers/review.mapper.ts
import type { ReviewDto } from '@gamescore/types';

import { toPlatformDto } from '../../games/mappers/game.mapper';
import { toUserSummary } from '../../users/mappers/user.mapper';
import type { ReviewWithRelations } from '../repositories/review.repository';

export function toReviewDto(
  review: ReviewWithRelations,
  extras: {
    viewerId?: string | null;
    viewerVote?: boolean | null;
    viewerHasReported?: boolean;
  } = {},
): ReviewDto {
  return {
    id: review.id,
    author: toUserSummary(review.user),
    game: review.game,
    recommended: review.recommended,
    rating: review.rating,
    text: review.text,
    hoursPlayed: review.hoursPlayed,
    platform: review.platform ? toPlatformDto(review.platform) : null,
    usefulCount: review.usefulCount,
    notUsefulCount: review.notUsefulCount,
    createdAt: review.createdAt.toISOString(),
    updatedAt: review.updatedAt.toISOString(),
    edited: review.edited,
    status: review.status,
    moderationStatus: review.moderationStatus,
    viewerVote:
      extras.viewerVote === true ? 'USEFUL' : extras.viewerVote === false ? 'NOT_USEFUL' : null,
    viewerIsAuthor: extras.viewerId === review.userId,
    viewerHasReported: extras.viewerHasReported ?? false,
  };
}

File: apps/api/src/GameScore.Infrastructure/Mappers/ReviewMapper.cs
using GameScore.Infrastructure.Mappers;
using GameScore.Infrastructure.Persistence;

namespace GameScore.Infrastructure.Mappers;

public static class ReviewMapper
{
    public static object ToDto(
        ReviewEntity review,
        UserEntity author,
        GameEntity game,
        PlatformEntity? platform,
        Guid? viewerId = null,
        bool? viewerVote = null,
        bool viewerHasReported = false) =>
        new
        {
            id = review.Id,
            author = UserMapper.ToSummary(author),
            game = new
            {
                id = game.Id,
                slug = game.Slug,
                name = game.Name,
                coverImageUrl = game.CoverImageUrl,
            },
            recommended = review.Recommended,
            rating = review.Rating,
            text = review.Text,
            hoursPlayed = review.HoursPlayed,
            platform = platform is null ? null : UserMapper.ToPlatform(platform),
            usefulCount = review.UsefulCount,
            notUsefulCount = review.NotUsefulCount,
            createdAt = review.CreatedAt.ToString("O"),
            updatedAt = review.UpdatedAt.ToString("O"),
            edited = review.Edited,
            status = review.Status.ToString(),
            moderationStatus = review.ModerationStatus.ToString(),
            viewerVote = viewerVote switch
            {
                true => "USEFUL",
                false => "NOT_USEFUL",
                _ => null,
            },
            viewerIsAuthor = viewerId == review.UserId,
            viewerHasReported,
        };
}
